src/controllers/dynakube/version/version.go

```diff
@@ -35,53 +35,90 @@ type Reconciler struct {
 	TimeProvider    *kubeobjects.TimeProvider
 }
 
-type updateSpec struct {
-	updateActiveGate bool
-	oneAgentUpdate   bool
+type updateScope struct {
+	onActiveGate bool
+	onOneAgent   bool
+	onSynthetic  bool
 }
 
-func (s updateSpec) needsUpdate() bool {
-	return s.updateActiveGate || s.oneAgentUpdate
+func (scope updateScope) needsUpdate() bool {
+	return scope.onActiveGate ||
+		scope.onOneAgent ||
+		scope.onSynthetic
+}
+
+func (scope updateScope) reconcileActiveGate(dynakube *dynatracev1beta1.DynaKube, updater *imageUpdater) {
+	if !scope.onActiveGate {
+		return
+	}
+
+	err := updater.update(
+		dynakube.ActiveGateImage(),
+		&dynakube.Status.ActiveGate.VersionStatus,
+		true)
+	if err != nil {
+		log.Error(err, "failed to update ActiveGate image version")
+	}
+}
+
+func (scope updateScope) reconcileOneAgent(dynakube *dynatracev1beta1.DynaKube, updater *imageUpdater) {
+	if !scope.onOneAgent {
+		return
+	}
+
+	err := updater.update(
+		dynakube.OneAgentImage(),
+		&dynakube.Status.OneAgent.VersionStatus,
+		false)
+	if err != nil {
+		log.Error(err, "failed to update OneAgent image version")
+	}
+}
+
+func (scope updateScope) reconcileSynthetic(dynaKube *dynatracev1beta1.DynaKube, updater *imageUpdater) {
+	if !scope.onSynthetic {
+		return
+	}
+
+	err := updater.update(
+		dynaKube.SyntheticImage(),
+		&dynaKube.Status.Synthetic.VersionStatus,
+		true)
+	if err != nil {
+		log.Error(err, "failed to update synthetic image version")
+	}
 }
 
 // Reconcile updates the version and hash for the images used by the rec.Dynakube DynaKube instance.
 func (r *Reconciler) Reconcile(ctx context.Context) error {
-	updateSpec := updateSpec{
-		updateActiveGate: needsActiveGateUpdate(r.Dynakube, *r.TimeProvider),
-		oneAgentUpdate:   needsOneAgentUpdate(r.Dynakube, *r.TimeProvider),
+	scope := updateScope{
+		onActiveGate: needsActiveGateUpdate(r.Dynakube, *r.TimeProvider),
+		onOneAgent:   needsOneAgentUpdate(r.Dynakube, *r.TimeProvider),
+		onSynthetic:  needsSynMonitoringUpdate(r.Dynakube, *r.TimeProvider),
 	}
 
-	if !updateSpec.needsUpdate() {
+	if !scope.needsUpdate() {
 		return nil
 	}
 
-	return r.updateImages(ctx, updateSpec)
+	return r.updateImages(ctx, scope)
 }
 
-func (r *Reconciler) updateImages(ctx context.Context, spec updateSpec) error {
+func (r *Reconciler) updateImages(ctx context.Context, scope updateScope) error {
 	dockerConfig, err := createDockerConfigWithCustomCAs(ctx, r.Dynakube, r.ApiReader, r.Fs)
 	if err != nil {
 		return err
 	}
 
-	imageUpdater := imageUpdater{
+	imageUpdater := &imageUpdater{
 		now:         *r.TimeProvider.Now(),
 		dockerCfg:   dockerConfig,
 		verProvider: r.VersionProvider,
 	}
-	if spec.updateActiveGate {
-		err := imageUpdater.update(r.Dynakube.ActiveGateImage(), &r.Dynakube.Status.ActiveGate.VersionStatus, true)
-		if err != nil {
-			log.Error(err, "failed to update ActiveGate image version")
-		}
-	}
+	scope.reconcileActiveGate(r.Dynakube, imageUpdater)
```

Comment:
revert this refactor, it's just code duplication in different methods.

Reply:
Yes, it's not perfect but such an arrangement reduce the cyclomatic index.

The linter pointed the `updateImages` as too complex when _statsd_ and _extension controller_ were being probed.